File: packages/python/src/axene_mailer/resources/templates.py
"""The ``templates`` resource: reusable email templates. Starter plan and up."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from .._http import HttpTransport
from .._serialize import prune


class Templates:
    """Accessed as ``axene.templates``."""

    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def list(self) -> List[Dict[str, Any]]:
        """List all templates, most recently updated first."""
        return self._http.request("GET", "/v1/templates/")

    def create(
        self,
        name: str,
        subject: Optional[str] = None,
        html: Optional[str] = None,
        text: Optional[str] = None,
        blocks_json: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Create a template.

        ``html`` maps to ``html_body`` and ``text`` to ``text_body`` on the
        wire. ``variables`` are derived server-side from ``{{name}}``
        placeholders, so you do not pass them.
        """
        body = prune(
            {
                "name": name,
                "subject": subject,
                "html_body": html,
                "text_body": text,
                "blocks_json": blocks_json,
            }
        )
        return self._http.request("POST", "/v1/templates/", body)

    def get(self, template_id: str) -> Dict[str, Any]:
        """Fetch a single template."""
        return self._http.request("GET", f"/v1/templates/{template_id}")

    def update(
        self,
        template_id: str,
        name: Optional[str] = None,
        subject: Optional[str] = None,
        html: Optional[str] = None,
        text: Optional[str] = None,
        blocks_json: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Update a template (partial). ``html``/``text`` map to
        ``html_body``/``text_body``.
        """
        body = prune(
            {
                "name": name,
                "subject": subject,
                "html_body": html,
                "text_body": text,
                "blocks_json": blocks_json,
            }
        )
        return self._http.request("PATCH", f"/v1/templates/{template_id}", body)

    def delete(self, template_id: str) -> None:
        """Delete a template."""
        return self._http.request("DELETE", f"/v1/templates/{template_id}")

    def duplicate(self, template_id: str) -> Dict[str, Any]:
        """Duplicate a template (the copy's ``blocks_json`` is not carried over)."""
        return self._http.request("POST", f"/v1/templates/{template_id}/duplicate")

File: packages/ruby/lib/axene/mailer/resources/templates.rb
# frozen_string_literal: true

module Axene
  module Mailer
    module Resources
      # The +templates+ resource: reusable email templates (Starter plan and up).
      # Accessed as +client.templates+.
      #
      # Note the wire mapping for this resource: +html+ maps to +html_body+ and
      # +text+ maps to +text_body+ (the emails resource keeps +html+/+text+).
      class Templates
        # @param transport [Axene::Mailer::Transport]
        def initialize(transport)
          @transport = transport
        end

        # List all templates, most recently updated first.
        #
        # @return [Array<Hash>]
        def list
          @transport.request(:get, "/v1/templates/")
        end

        # Create a template. +variables+ are derived server-side from {{name}}
        # placeholders in the bodies, so you do not pass them.
        #
        # @param name [String]
        # @param subject [String, nil]
        # @param html [String, nil] maps to html_body
        # @param text [String, nil] maps to text_body
        # @param blocks_json [Hash, nil]
        # @return [Hash]
        def create(name:, subject: nil, html: nil, text: nil, blocks_json: nil)
          @transport.request(:post, "/v1/templates/", body: serialize(name, subject, html, text, blocks_json))
        end

        # Fetch a single template.
        #
        # @param id [String]
        # @return [Hash]
        def get(id)
          @transport.request(:get, "/v1/templates/#{Util.escape(id)}")
        end

        # Update a template (partial).
        #
        # @param id [String]
        # @param name [String, nil]
        # @param subject [String, nil]
        # @param html [String, nil] maps to html_body
        # @param text [String, nil] maps to text_body
        # @param blocks_json [Hash, nil]
        # @return [Hash]
        def update(id, name: nil, subject: nil, html: nil, text: nil, blocks_json: nil)
          @transport.request(:patch, "/v1/templates/#{Util.escape(id)}",
                             body: serialize(name, subject, html, text, blocks_json))
        end

        # Delete a template.
        #
        # @param id [String]
        # @return [nil]
        def delete(id)
          @transport.request(:delete, "/v1/templates/#{Util.escape(id)}")
        end

        # Duplicate a template (the copy's +blocks_json+ is not carried over).
        #
        # @param id [String]
        # @return [Hash]
        def duplicate(id)
          @transport.request(:post, "/v1/templates/#{Util.escape(id)}/duplicate")
        end

        private

        def serialize(name, subject, html, text, blocks_json)
          Util.prune(
            name: name,
            subject: subject,
            html_body: html,
            text_body: text,
            blocks_json: blocks_json
          )
        end
      end
    end
  end
end
